linera-chain/src/chain.rs

```diff
@@ -625,30 +612,56 @@ where
                 },
                 authenticated_signer: message.event.authenticated_signer,
             };
-            let results = self
-                .execution_state
-                .execute_message(
-                    context,
-                    message.event.message.clone(),
-                    &policy,
-                    &mut tracker,
-                )
-                .await
-                .map_err(|err| ChainError::ExecutionError(err, chain_execution_context))?;
+            let results = match message.action {
+                MessageAction::Accept => self
+                    .execution_state
+                    .execute_message(
+                        context,
+                        message.event.message.clone(),
+                        &policy,
+                        &mut tracker,
+                    )
+                    .await
+                    .map_err(|err| ChainError::ExecutionError(err, chain_execution_context))?,
+                MessageAction::Reject => {
+                    ensure!(
+                        !message.event.is_protected(),
+                        ChainError::CannotRejectMessage {
+                            chain_id,
+                            origin: Box::new(message.origin.clone()),
+                            event: message.event.clone(),
+                        }
+                    );
+                    if message.event.is_tracked() {
+                        // Bounce the message.
+                        self.execution_state
+                            .bounce_message(context, message.event.message.clone())
+                            .await
+                            .map_err(|err| {
+                                ChainError::ExecutionError(err, chain_execution_context)
+                            })?
+                    } else {
+                        // Nothing to do.
+                        Vec::new()
+                    }
+                }
+            };
             let mut messages_out = self
                 .process_execution_results(context.height, results)
                 .await?;
-            let balance = self.execution_state.system.balance.get_mut();
-            Self::sub_assign_fees(
-                balance,
-                policy.storage_bytes_written_price_raw(&message)?,
-                chain_execution_context,
-            )?;
-            Self::sub_assign_fees(
-                balance,
-                policy.messages_price(&messages_out)?,
-                chain_execution_context,
-            )?;
+            if let MessageAction::Accept = message.action {
```

Comment:
(I guess when we have messages that pay their own fees, we'll also account for the bouncing cost in the `Reject` case?)